test/pytest/service-desktop/test_contacts.py

```diff
@@ -1,146 +1,357 @@
-
-    # gathering updated contact
-    body = {"id": contact_id_to_update}
-    ret = harness.endpoint_request("contacts", "get", body)
-    contact = {"address": "6 Czeczota St.\n02600 Warsaw",
-               "altName": "Testowy2",
-               "email": "testowy.2@example.com",
-               "blocked": False,
-               "favourite": True,
-               "ice": True,
-               "numbers": ["547623521"],
-               "speedDial": "7",
-               "priName": "Test2",
-               "note": "this is a really cool guy",
-               "id": contact_id_to_update}
-    assert ret["body"] == contact
-
-    # removing added contact
-    body = {"id": contact_id_to_update}
-    ret = harness.endpoint_request("contacts", "del", body)
-    assert ret["status"] == status["NoContent"]
-
-    # verifying count
-    body = {"count": True}
-    ret = harness.endpoint_request("contacts", "get", body)
-    assert ret["status"] == status["OK"]
-
-    assert ret["body"]["count"] == count
+def test_edit_contact(harness):
+    # Check initial state
+    contacts_tester = ContactsTester(harness)
+    result, initial_number_of_contact_records = contacts_tester.get_contacts_count()
+    assert result, "Failed to get contacts count!"
+
+    # Add contact
+    contact_record = {
+        "address": "6 Czeczota St.\n02600 Warsaw",
+        "altName": "Smith",
+        "email": "john.smith@mudita.com",
+        "blocked": False,
+        "favourite": False,
+        "ice": False,
+        "numbers": [
+            "123456789"
+        ],
+        "speedDial": "1",
+        "priName": "John",
+        "note": "Some note"
+    }
+
+    result, contact_id = contacts_tester.add_contact(contact_record)
+    assert result, "Failed to add contact!"
+
+    # Edit contact
+    edited_contact_record = {
+        "id": contact_id,
+        "address": "6 Czeczota St.\n02600 Warsaw",
+        "altName": "Davies",
+        "email": "john.davies@mudita.com",
+        "blocked": False,
+        "favourite": False,
+        "ice": False,
+        "numbers": [
+            "123456799"
+        ],
+        "speedDial": "1",
+        "priName": "John",
+        "note": "Some note"
+    }
+    assert contacts_tester.update_contact(edited_contact_record), "Failed to update contact!"
+    result, retrieved_contact_record = contacts_tester.get_contact_by_id(contact_id)
+    assert result, "Failed to get contact by ID!"
+    assert retrieved_contact_record == edited_contact_record
+
+    # Clean up
+    assert contacts_tester.delete_contact_by_id(contact_id), "Failed to delete a contact!"
+    result, received_contact_records_count = contacts_tester.get_contacts_count()
+    assert result, "Failed to get contacts count!"
+    assert initial_number_of_contact_records == received_contact_records_count
+
+
+@pytest.mark.service_desktop_test
+@pytest.mark.usefixtures("phone_unlocked")
+def test_add_and_get_multiple_contacts(harness):
```

Comment:
This test is very long, which in itself is nothing wrong but the amount of possible failure points (see the amount of asserts for example) is so big, that maintenance of this test will be a nightmare. Imagine it fails, and you see it on the report. What information are you getting at this point? Not very informative since any of those multiple asserts could've been the reason.

Reply:
I agree that this test is very long.
Unfortunately, it seems it has to be like this at this moment.
I could change this test to 4 separate tests: 
1. test adding multiple contacts
2. test getting multiple contacts
3. test limit
4. test offset.
But then, to test 4, I need to copy almost all code from the previous test to it.
Same for 3, almost all code from tests 1 and 2 should be copied to it.
Overall, we end up with 4 tests instead of one, but two of them are almost that big, like the one we divided.

However, I see one solution for this.
If we could have prepared DB files with proper test data, we could simplify this (and not only) test by a lot.
We could utilize a backup and restore mechanism for this. We could also address the clean-up problem with this method.

This potential improvement should be considered in the next iteration of improving this, and other tests.